File: leetcode/46.全排列.go
func permute(nums []int) [][]int {
    ans := [][]int{}
    path := []int{}
    vis := make([]bool, len(nums))
    var dfs func()
    dfs = func() {
        if len(path) == len(nums) {
            ans = append(ans, append([]int(nil), path...))
            return
        }
        for i := 0; i < len(nums); i++ {
            if vis[i] {
                continue
            }
            path = append(path, nums[i])
            vis[i] = true
            dfs()
            vis[i] = false
            path = path[: len(path) - 1]
        }
    }
    dfs()
    return ans
}

File: leetcode/46.全排列.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []
        t = []
        n = len(nums)
        vis = [False] * n
        def dfs(i: int):
            if i == n:
                ans.append(t[::])
                return
            for j in range(n):
                if not vis[j]:
                    t.append(nums[j])
                    vis[j] = True
                    dfs(i+1)
                    vis[j] = False
                    t.pop() 
        dfs(0)
        return ans